openfl/component/straggler_handling_functions/cutoff_time_based_straggler_handling.py

```diff
@@ -30,11 +31,100 @@ def __init__(
         self.straggler_cutoff_time = straggler_cutoff_time
         self.minimum_reporting = minimum_reporting
+        self.logger = getLogger(__name__)
+
+        if self.straggler_cutoff_time == np.inf:
+            self.logger.warning(
+                "CutoffTimeBasedStragglerHandling is disabled as straggler_cutoff_time "
+                "is set to np.inf."
+            )
+
+    def reset_policy_for_round(self) -> None:
+        """
+        Reset timer for the next round.
+        """
+        if hasattr(self, "timer"):
+            self.timer.cancel()
+            delattr(self, "timer")
+
+    def start_policy(self, callback: Callable) -> None:
+        """
+        Start time-based straggler handling policy for collaborator for
+        a particular round.
 
-    def straggler_time_expired(self):
+        Args:
+            callback: Callable
+                Callback function for when straggler_cutoff_time elapses
+
+        Returns:
+            None
+        """
+        # If straggler_cutoff_time is set to infinity
+        # or if the timer is already running,
+        # do not start the policy.
+        if self.straggler_cutoff_time == np.inf or hasattr(self, "timer"):
+            return
+        self.round_start_time = time.time()
+        self.timer = threading.Timer(
+            self.straggler_cutoff_time,
+            callback,
+        )
+        self.timer.daemon = True
+        self.timer.start()
+
+    def straggler_cutoff_check(
+        self,
+        num_collaborators_done: int,
+        num_all_collaborators: int,
+    ) -> bool:
+        """
+        If minimum_reporting collaborators have reported results within
+        straggler_cutoff_time then return True, otherwise False.
+
+        Args:
+            num_collaborators_done: int
+                Number of collaborators finished.
+            num_all_collaborators: int
+                Total number of collaborators.
+
+        Returns:
+            bool
+        """
+
+        # if straggler time has not expired then
+        # wait for ALL collaborators to report results.
+        if not self.__straggler_time_expired():
+            return num_all_collaborators == num_collaborators_done
+
+        # Check if time has expired
+        elif self.__straggler_time_expired():
+            # Check if minimum_reporting collaborators have reported results
+            if self.__minimum_collaborators_reported(num_collaborators_done):
+                self.logger.info(
+                    f"{num_collaborators_done} collaborators reported results within "
+                    "cutoff time. Applying cutoff policy and proceeding with end of round."
+                )
+                return True
+            self.logger.info(
+                "Disregarded straggler handling policy and waiting for minimum "
+                f"{self.minimum_reporting} collaborator(s) to report results."
+            )
+            return False
```

Comment:
I suggest to put this case under an explicit `else:` block

Reply:
Updated. Thanks!